base/abstractarraymath.jl

```diff
@@ -386,6 +386,34 @@ end
 
 end#module
 
+struct EachSlice{E,M,A,I,L} <: AbstractArray{E,M}
+    "parent array"
+    arr::A
+    "The `CartesianIndices` iterator used to index each slice"
+    cartiter::I
+    """
+    A `Tuple` of length `ndims(A)`, each element either the dimension index into
+    `cartiter`, or `nothing` if it is a variable to be sliced over
+    """
+    lookup::L
+end
+
+function iterate(s::EachSlice, state...)
+    r = iterate(s.cartiter, state...)
+    r === nothing && return r
+    (c,nextstate) = r
+    view(s.arr, map(l -> l === nothing ? (:) : c[l], s.lookup)...), nextstate
```

Comment:
Ah, and `view` isn't guaranteed to return a `SubArray` — we should probably call into the constructor directly.

Reply:
I was wondering about this (I know it's not true for e.g. `CuArray`s). Would it make sense to go the other way, i.e. have a function that returns the type of whatever the `view` would be?

Reply:
Yeah, we could do that.  I think I'd lean towards doing it in the value domain, though, as then we can fall back to a simple `typeof(view(...))`.  The whole point of views is that they should be cheap (or even no-cost) to construct (especially if not used).

Reply:
The problem here is in the empty case - we still need to return something.